### PR TITLE
Add io.github.sharkwouter.Minigalaxy

### PR DESCRIPTION
**Pull request by tim77**
Hello. This is my first Flatpak package and i expect upstream involvement here and transfer ownership when upstream [will be ready/have time](https://github.com/sharkwouter/minigalaxy/issues/11) for this.

Such kind of apps is problematic to package via Flatpak format and it will have some limitation probably, but in same time there is almost no alternative available on Flathub, so it would be nice to have some on Flathub.

CC: @sharkwouter
Also thanks @gasinvein for help with packaging.

### CODE REVIEW

**Comment by sharkwouter** [comment]
I would be willing to take ownership of this, if it ends up working well. The main issue will most likely be launching games from within Minigalaxy. Which makes it very hard to get working. Another attempt was made in [this repo](https://github.com/protheory8/minigalaxy-flatpak) by @protheory8 but we had issues making games able to launch from it.

A possible solution for a lot of games would be to include Dosbox and Scummvm, but that would probably add a ton of dependencies and increase complexity. I think it would be hard to get games which use those tools to launch without them, though.

The problem addressed with the patch has been addressed upstream as well, but it didn't make it into the 0.9.3 release. You can find the commit at https://github.com/sharkwouter/minigalaxy/commit/b0f71e5c05972314bf5a7318019e600b2b8d9daf, maybe it is of use.

**Comment by gasinvein** [comment]
> A possible solution for a lot of games would be to include Dosbox and Scummvm, but that would probably add a ton of dependencies and increase complexity. I think it would be hard to get games which use those tools to launch without them, though.

Both of them are already on Flathub, but as standalone apps. Probably they could be reused as Minigalaxy flatpak extensions in the future?

File: com.github.sharkwouter.minigalaxy.yaml [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+    no-autodownload: true
+    autodelete: false
+    add-ld-path: lib
+    merge-dirs: vulkan/icd.d;glvnd/egl_vendor.d;OpenCL/vendors;lib/dri;lib/d3d;vulkan/explicit_layer.d
+    download-if: active-gl-driver
+    enable-if: active-gl-driver
+
+command: minigalaxy
+rename-desktop-file: minigalaxy.desktop
+rename-icon: minigalaxy
```
Comment (p1u3o) [inline com.github.sharkwouter.minigalaxy.yaml]: Might be an idea to eventually try and get these renamed upstream too.

**Comment by p1u3o** [comment]
bot, build com.github.sharkwouter.minigalaxy

File: com.github.sharkwouter.minigalaxy.yaml (hunk 1 above)
Comment (tim77) [reply com.github.sharkwouter.minigalaxy.yaml]: Also AppData check not passed. Seems like we need to add additional icon size formats to satisfy linter. I don't mind send PR if @sharkwouter will approve this initiative.

Comment (sharkwouter) [reply com.github.sharkwouter.minigalaxy.yaml]: I would. If it helps, the original source for the icon is one of the jump images in this set: https://opengameart.org/content/jumping-galaxy-asset-cc-by-30

Comment (sharkwouter) [reply com.github.sharkwouter.minigalaxy.yaml]: @p1u3o what should the image and desktop file be called for these lines to not be needed?

Comment (tim77) [reply com.github.sharkwouter.minigalaxy.yaml]: > what should the image and desktop file be called for these lines to not be needed?

Should use [reverse-DNS scheme](https://www.freedesktop.org/software/appstream/docs/chap-Quickstart.html#sect-Quickstart-DesktopApps). Like
`com.github.sharkwouter.minigalaxy.png`
`com.github.sharkwouter.minigalaxy.desktop`

File: com.github.sharkwouter.minigalaxy.yaml [hunk 2]
```diff
@@ -0,0 +1,61 @@
+---
+app-id: com.github.sharkwouter.minigalaxy
+runtime: org.gnome.Platform
+runtime-version: '3.34'
```
Comment (LeandroStanger) [inline com.github.sharkwouter.minigalaxy.yaml]: Update runtime to 3.36

**Comment by LeandroStanger** [comment]
bot, build com.github.sharkwouter.minigalaxy

File: com.github.sharkwouter.minigalaxy.yaml (hunk 1 above)
Comment (sharkwouter) [reply com.github.sharkwouter.minigalaxy.yaml]: I made some changes upstream now. The desktop, icon and metainfo files have been renamed to use the ``com.github.sharkwouter.minigalaxy`` naming scheme. Both the 128x128 and 192x192 icons are now included.

Comment (sharkwouter) [reply com.github.sharkwouter.minigalaxy.yaml]: Should I create a tag with these changes to make it easier to build for?

Comment (tim77) [reply com.github.sharkwouter.minigalaxy.yaml]: > Should I create a tag with these changes to make it easier to build for?

Would be great.

Comment (sharkwouter) [reply com.github.sharkwouter.minigalaxy.yaml]: Try that flatpak-1 tag

**Comment by tim77** [comment]
bot, build com.github.sharkwouter.minigalaxy

**Comment by tim77** [comment]
bot, build com.github.sharkwouter.minigalaxy

**Comment by sharkwouter** [comment]
I did a quick test  by building this on my system. It seems no games are able to launch from it. With native ones the launch script files to launch the game. I think the gog bash library is not able to change the working directory, but I'm not sure why.

**Comment by gasinvein** [comment]
@sharkwouter You probably need to install `org.gnome.Platform.Compat.i386//3.36` manually. Flatpak doesn't automatically install app extensions if they come from different repositories than the app itself.

**Comment by barthalion** [comment]
@tim77 Please have a look at @gasinvein's comment.

**Comment by tim77** [comment]
> @tim77 Please have a look at @gasinvein's comment.

Many games won't run, this is true. But at least improved build a little bit. Still very WIP and i've contacted @gasinvein and hope we will improve tomorrow build...

**Comment by LeandroStanger** [comment]
bot, build com.github.sharkwouter.minigalaxy

**Comment by AsciiWolf** [comment]
> Build [17355 failed](https://flathub.org/builds/#/builders/32/builds/17355)

[OARS](https://hughsie.github.io/oars/) metadata seem to be missing.

**Comment by sharkwouter** [comment]
I have no idea how to fill out this information. Minigalaxy doesn't include anything offensive, but you can download some "adult" content from GOG

**Comment by tim77** [comment]
> I have no idea how to fill out this information. Minigalaxy doesn't include anything offensive, but you can download some "adult" content from GOG

Yeah, i've been thinking about this as well and not 100% sure but we have great example -- Steam. I think we should do the same for _Minigalaxy_ and raise at max/moderate alsmost all this OARS stuff:
Steam OARS

I've sent the initial PR for OARS stuff.

**Comment by sharkwouter** [comment]
@tim77 I've made a tag called flatpak-3 which includes the OARS information

**Comment by tim77** [comment]
bot, build com.github.sharkwouter.minigalaxy

**Comment by sharkwouter** [comment]
Well, it did succeed this time

**Comment by LeandroStanger** [comment]
bot, build com.github.sharkwouter.minigalaxy

**Comment by LeandroStanger** [comment]
bot, build com.github.sharkwouter.minigalaxy

**Comment by LeandroStanger** [comment]
Captura de tela de 2020-03-31 11-19-00

**Comment by sharkwouter** [comment]
That is to be expected for Dosbox games. The version delivered by GOG is not compatible with anything but Ubuntu for afaik.

**Comment by tim77** [comment]
This `libSDL_sound-1.0.so.1` not symlink, just a text file with contain `libSDL_sound-1.0.so.1.0.2`. But most games now runs fine which i tested. Also we can add in future this https://github.com/flathub/flathub/pull/1428 and continue to extent base apps for game launchers.

**Comment by protheory8** [comment]
I actually think we can probably use [this](https://manpages.debian.org/experimental/flatpak/flatpak-spawn.1.en.html) thing something

**Comment by protheory8** [comment]
I would recommend renaming it to `io.github.sharkwouter.Minigalaxy` because:
1. https://sharkwouter.github.io/minigalaxy exists.
2. Flatpak Docs recommend using `io.github` instead of `com.github` [here](https://docs.flatpak.org/en/latest/conventions.html#application-ids).
3. `Minigalaxy` starts from capital letter because that's what used in Flatpak Docs and a lot of apps do the same thing, [example](https://flathub.org/apps/details/org.blender.Blender).

**Comment by sharkwouter** [comment]
I've added a new tag called flatpak-4 includes the changes which were just made to improve the flatpak experience.

**Comment by tim77** [comment]
Tested `flatpak-4` build and now we struggle with another issue (same `Bio Menace` example):

```
Executing integrity check for /home/tim/.var/app/io.github.sharkwouter.Minigalaxy/cache/minigalaxy/download/Bio Menace
Verifying archive integrity... MD5 checksums are OK. All good.
warning [/home/tim/.var/app/io.github.sharkwouter.Minigalaxy/cache/minigalaxy/download/Bio Menace]:  815817 extra bytes at beginning or within zipfile
  (attempting to process anyway)
checkdir:  cannot create extraction directory: /home/tim/.var/app/io.github.sharkwouter.Minigalaxy/cache/minigalaxy/extract/1449569170
           No such file or directory
```

**Comment by sharkwouter** [comment]
I see, try flatpak-5

**Comment by tim77** [comment]
bot, build io.github.sharkwouter.Minigalaxy

**Comment by tim77** [comment]
Well, almost all games which i tested from this list [1] works great for me, except some, like Trine, Wastelands 2, but they not run either on `minigalaxy` non-flatpak version, so this make me believe that this issues not flatpak related. And i'll continue to reporting bugs to `minigalaxy` if you don't mind (i have few).

Note: additional help with testing games and reporting would be very much helpful.

[1]:
Снимок экрана от 2020-04-06 20-43-40

**Comment by sharkwouter** [comment]
Great, keep up the good work!

**Comment by tim77** [comment]
Now i can say that literally every game which i have tested running in _minigalaxy_ flatpak build. My problem with _Wastelands 2_ was with `xdg-desktop-portal` itself when i changed my location for games. This has nothing to do with flatpak build and this is a new feature. I should report bug in `xdg-desktop-portal` but i don't know how to file a bug and report this properly at this time.

So, games runs fine without issues. For _Trine_ we need `gtk2` here. @gasinvein bundled `gtk2` libs in his [GameLauncherBaseApp](https://github.com/gasinvein/flatpak-apps-staging/blob/org.flathub.GameLauncherBaseApp/master/modules/gtk2-libs.yml) so we can bundle them too in _minigalaxy_ or we can use in future this GameLauncherBaseApp. At this time Flatpak daddies [sceptical](https://github.com/flathub/flathub/pull/1428#issuecomment-610090870) about this idea in general. In my opinion this can help a lot and reduce maintenance burden for all such kind of apps like minigalaxy, GameHub, Lutris.

**Comment by sharkwouter** [comment]
Not sure if it helps, but the next release of Minigalaxy is planned to happen on the 20th of April and will include the changes made for the flatpak package.

It also ships with support for downloading and installing Windows games with wine if it is installed, but it shouldn't break anything on systems which don't have it installed. Not sure if that's worth noting, since it's probably not worthwhile to try to implement support for it in the first release of the flatpak package.

**Comment by tim77** [comment]
Now since `0.9.4` released the question about WINE is open. I still believe we need [this](https://github.com/gasinvein/flatpak-apps-staging/tree/org.flathub.GameLauncherBaseApp/master) to solve this.

**Comment by LeandroStanger** [comment]
bot, build io.github.sharkwouter.Minigalaxy

**Comment by barthalion** [comment]
bot, build io.github.sharkwouter.Minigalaxy

**Comment by barthalion** [comment]
Let's bundle gtk2 here for the time being. Other than this, LGTM.

**Comment by tim77** [comment]
Minigalaxy Flatpak build works OK even at this moment but with some limited functionality for sure. Like WINE, gtk2, and probably tons some legacy libs which we need to bundle but we never know which exactly until we massively test it and receive some feedback from users. So we need to decide are we afraid bugs or ready for pushing it to Flathub repos? :)

Also, where global WINE discussion about how it should be packaged on Flathub? And what version we should ship? Staging, Stable, TGK? How WINE should packaged... There many questions. Staging actual version i found in @gasinvein [manifests](https://github.com/gasinvein/flatpak-apps-staging/blob/org.winehq.Wine/master/org.winehq.Wine.yml).

**Comment by barthalion** [comment]
It's up to you. We can merge the application to beta branch and once you verify it does what it should (be it via mentioning on project's website and getting more users on board), you're free to push it to `master` to publish to regular Flathub.

How to bite Wine is out of scope for this PR. It requires more discussion and actual design before we start pumping new extensions/applications/whatever into Flathub.

**Comment by tim77** [comment]
bot, build io.github.sharkwouter.Minigalaxy

**Comment by tim77** [comment]
So, let's merge it into Stable repo? BTW "Trine" still won't run with bundled `gtk2` but now because of `libz`. _Trine_ is [special](https://www.gog.com/forum/trine_series/library_incompatibility_breaks_trine_on_linux) case so hope we investigate this in future.

**Comment by tim77** [comment]
The reason is why i want push it Stable Flathub repo is simple: i don't even know until recently that _Lutris_ available in Flathub Beta repo. So it will almost undiscoverable if in Beta repos, but if upstream mention this at least somewhere then maybe fine.

**Pull request closed without merging**